CRM/Civisocial/OAuthProvider.php

```diff
@@ -160,31 +170,79 @@ public function handleCallback() {
   }
 
   /**
+   * Save social user to the database
+   *
+   * @param string $oAuthProvider
+   *   Shortname for OAuth provider
+   * @param array $userProfile
+   *   Social user information
+   */
+  public function saveSocialUser($oAuthProvider, $userProfile) {
+    $session = CRM_Core_Session::singleton();
+
+    $socialUserId = CRM_Utils_Array::value("id", $userProfile);
+    $contactId = civicrm_api3(
+      'CivisocialUser',
+      'socialuserexists',
+      array(
+        'social_user_id' => $socialUserId,
+        'oauth_provider' => $oAuthProvider,
+      )
+    );
+
+    if (!$contactId) {
+      $user = array(
+        'first_name' => CRM_Utils_Array::value('first_name', $userProfile),
+        'last_name' => CRM_Utils_Array::value('last_name', $userProfile),
+        'display_name' => CRM_Utils_Array::value("name", $userProfile),
+        'preffered_language' => CRM_Utils_Array::value("locale", $userProfile),
+        'gender' => CRM_Utils_Array::value('gender', $userProfile),
+        'email' => CRM_Utils_Array::value("email", $userProfile),
+        'contact_type' => 'Individual',
+      );
+
+      // Find/create contact to map with social user
+      $contactId = civicrm_api3('CivisocialUser', 'createcontact', $user);
+
+      // Create social user
+      $socialUser = array(
+        'contact_id' => $contactId,
+        'social_user_id' => $socialUserId,
+        'access_token' => serialize($session->get('access_token')),
+        'oauth_provider' => $oAuthProvider,
+        'created_date' => time(), // @todo: Created Date not being recorded
+      );
+
+      civicrm_api3('CivisocialUser', 'create', $socialUser);
+    }
+    $this->login($oAuthProvider, $socialUserId, $contactId);
+    $this->redirect();
+  }
+
+  /**
    * Save OAuth Provider information to the session.
    * Acts as a logout if no parameters is passed.
    *
    * @param string $OAuthProvider
    *   Shortname for OAuth provider
-   * @param string $accessToken
-   *   Access Token provided by OAuth provider after successfull authentication
    * @param string $OAuthProviderId
    *   Unique user ID to OAuthProvider
    * @param int $contactId
    *   Contact ID of the social user
    *
    */
-  public function login($OAuthProvider = NULL, $accessToken = NULL, $OAuthProviderId = NULL, $contactId = NULL) {
+  public function login($OAuthProvider = NULL, $OAuthProviderId = NULL, $contactId = NULL) {
     $session = CRM_Core_Session::singleton();
 
-    if ($OAuthProvider == NULL && $accessToken == NULL && $OAuthProviderId == NULL && $contactId == NULL) {
+    if ($OAuthProvider == NULL && $OAuthProviderId == NULL && $contactId == NULL) {
```

Comment:
There should be defensive checks for existence of contactId. The above lines means having anyone of the above will set the stats as logged in.